spock-specs/src/test/groovy/org/spockframework/runtime/model/SpecInfoHierarchySpec.groovy

```diff
@@ -0,0 +1,94 @@
+package org.spockframework.runtime.model
+
+import spock.lang.Shared
+import spock.lang.Specification
+import spock.lang.Unroll
+
+class SpecInfoHierarchySpec extends Specification {
+
+  @Shared def grandparent = new SpecInfo()
+  @Shared def parent = new SpecInfo()
+  @Shared def specInfo = new SpecInfo()
+  @Shared def child = new SpecInfo()
+  @Shared def grandchild = new SpecInfo()
+
+  @Shared hierarchy = [grandparent, parent, specInfo, child, grandchild]
+  @Shared hierarchyNames = ["grandparent", "parent", "specInfo", "child", "grandchild"]
+
+  def setupSpec() {
+    parent.superSpec = grandparent
+    grandparent.subSpec = parent
+
+    specInfo.superSpec = parent
+    parent.subSpec = specInfo
+
+    child.superSpec = specInfo
+    specInfo.subSpec = child
+
+    grandchild.superSpec = child
+    child.subSpec = grandchild
+  }
```

Comment:
I will look into that. But actually, I wanted to clearly express the "family tree" here and nothing else, hence also the extra names instead of simply constructing a hierarchy using a loop or so. My goal was for the spec to really be a specification documenting a feature. You read it, and it is easy to see what is going on and what is being verified. I thought it was worth the verbosity in the test/spec, because the (corner) cases I fixed are not so obvious otherwise.

Reply:
I just looked into it and am failing to see the benefit. I even think it would be confusing to create 5 spec infos like that. Which classes should I use to parametrise the builder's constructor in the first place? There are not classes. I am just testing the hierarchy helper methods in isolation. My suggestion here is to not change anything, if that would be acceptable to you. It would just complicate and obscure the test.

Reply:
I was suggesting to actually create 5 Spec classes, with the correct relationship. You could either use standalone classes or, you could use the `EmbeddedSpecCompiler` to create those classes for you, the upside is that they don't appear anywhere else, the downside is that they can't be addressed as nicely. See org.spockframework.smoke.Blocks or 
```groovy
  def "Create Hierarchy"() {
    def classes = compiler.compile("""import spock.lang.Specification
class Grandparent extends  Specification {}
class Parent extends Grandparent {}
class Middle extends Parent {}
class Child extends Middle {}
class Grandchild extends Child {}
""")

    def byName = classes.collectEntries{ [(it.name): it] }
    expect:
    byName.keySet() =~ ["Grandparent", "Parent", "Middle", "Child", "Grandchild"]
  }
```

of course this would be in a static helper method returning the `byName` map.